### ISSUE TITLE
v3 Tracks

### ISSUE THREAD

**Issue by iHiD** [exercism/exercism#5088]
# V3 tracks

This is a working proposal for the next iteration of Exercism. I would welcome your feedback to understand any blindspots we might have, things we've missed or got wrong, and any challenges we may not have considered.

## Background

Exercism v1 (the original site) had no concept of progression or structure within a language. A track was a bucket of exercises, which users were free to choose from.

In designing v2, we added a structured pathway (core exercises) which aimed to take a user through the basis of what they needed to know for that language, then additional (side) exercises that allowed them to explore other related concepts. This worked great in theory, but a year later, despite a lot of work by maintainers, no tracks have core pathways that do a great job of comprehensively teaching the key concepts.

Why is this?

Firstly, we believe the key reason is that we have encouraged maintainers to reorder their existing tracks, rather than design their tracks from scratch. The original exercises were never designed to teach language concepts - they were just chosen as fun things to experiment with - so shoehorning these into a "curriculum" is a painful, and probably impossible, task for maintainers. Secondly, in our strategies in trying to achieve this, we have focused too much on what was similar about languages, and not looked enough at the key question of "what makes languages unique?"

## Current problems

We've been taking stock of this situation recently and spending time trying to work out what the content of an Exercism track should look like in order to fulfil the design goals of v2. 

At the moment the journey for a student is:
- Be given an exercise.
- Try something - normally via an incorrect approach grounded from previous language experience.
- Be told what you should have been doing.
- Work out how to redo it that way.

That sometimes works really well. The mentor gives feedback that neither gives away the answer, nor is frustratingly vague, the student realises a new approach, gets a dopamine hit, and memorizes the new technique. This is similar to how many learning natural-language courses (such as Rosetta Stone) work, where you are given a set of images, and some text, and you have to work out which is which. At the start you have no idea, but as you are corrected you start to work out what the words mean, and the resulting dopamine hit locks the knowledge into your mind. 

However, in reality on Exercism, this is often a frustrating process for students. For more complex exercises (I'm looking at you, Bob), the student will often have no idea of the idiomatic way to approach it, and so write a very different and unidiomatic implementation, giving a mentor a huge amount of work to take them on a journey to an idiomatic solution. This can be hugely frustrating for the student, who can't see the end goal enough to be able to understand why they're being taken on the journey, and is a huge piece of work for the mentors (indeed, tracks pretty reliably fail as soon as a complex exercise like Bob is introduced). There are also huge waits between submitting an exercise and receiving feedback, and that break of flow reduces a lot of the positive excitement from learning something. Everything also relies on the mentor knowing exactly what the exercise is aiming for and not taking the student down random roads that are not appropriate to the point in the track. We have tried to solve some of these issues with the automatic analyzers, but the large potential amount of solutions submitted to an exercise, makes this a slow and unwieldy development endeavour.

## Evolving tracks

In learning programming languages, it is essential to understand a concept, work out how to apply that concept, and then practice using that concept in different examples.

At the moment I don't feel like we get this right. We let a student experiment and get it wrong, then give some hints, then ask them to work it out from that and maybe receive more hints. I'm not convinced that at the end of an exercise a student really understands what they've learnt, or how to apply it in the future. This is validated by the fact that on later exercises students often make the same "mistakes" they made early on, or don't apply the techniques they've been introduced to.

I believe the key area where we are failing our students is in not having a clearer separation between explaining a concept and giving the student an opportunity to play. At no point does Exercism actually go about teaching the concepts that someone needs to learn a language - we rely on mentors giving the correct hints when a student makes a mistake. I believe we need to rework core and side exercises to achieve this. I propose renaming "core exercises" to Concept Exercises and structuring them so that they introduce a specific concept in the instructions and then challenge the student to use that concept in producing a solution; and renaming "side exercises" to Practice Exercises, which are designed to allow students to solve an arbitary problem, with the aim of them making use of the concepts they have learnt so far.

### Concept Exercises

Concepts Exercises should have the following characteristics:
- Each one has a clear learning goal.
- They are language-specific, not generic.
- We use stubs/boilerplate to avoid the student having to learn/handle unnecessary code on exercises.

As Concept Exercises will be smaller and contain boilerplate, they will be much easier to generate auto-analyzers for. By also adding [Representers](https://github.com/exercism/exercism/issues/5079) we believe we can achieve 95% auto-analysis of iterations. We will automatically post feedback both giving students tips when they submit an approvable solution, but also providing guidance where the iteration is not approvable. Furthermore, our analyzers will automatically approve any approvable iteration, removing the tension/frustration that is often caused by that power lying with mentors. For the small percentage of solutions that cannot be automatically analyzed, these should provide an interesting and unique experience to mentors.

We will also be changing some of the rules around unlocking. Multiple Concept Exercises will be able to be unlocked at the same time, and Concept Exercises can unlock more complex, related Concept Exercises (e.g. Ruby might have a blocks Concept Exercise, which when completed unlocks exercises on Procs, Lambdas, block_given?, etc).

We will also change the way unlocking works, so rather than being explicitly stated in track configurations, unlocking will be implicitly worked out by the tagging of "concepts learnt" on Concept Exercises, and "prerequisite concepts" on Practice Exercises. Once enough Concept Exercises have been completed that all the prerequisite concepts for a Practice Exercise has been been fulfilled, that Practice Exercise will be unlocked.

Concept Exercises will not share a generic base such as problem-specifications. However, we will write language-agnostic introductions to programming concepts, with links to language-generic resources on that topic, which we will ask tracks to use. For example, we might define a concept of "Enumeration" and provide a short introduction that explains how enumeration differs from loops with a link out to a good article (either internal to Exercism or elsewhere on the web). Tracks implementing an exercise on enumeration can then include this common paragraph, before explaining their own language-specific semantics and syntax.

### Practice Exercises

Practice Exercises should be story-based than concept-based. They should provide an interesting challenge, which a student can try and solve using the concepts they have been taught. The existing set of Exercism exercises are generally great Practice Exercise candidates.

There will be no "approval" for Practice Exercises. A student solves the exercise against the tests and submits it. They can then choose to publish, or submit further iterations. 

We will provide automated analysis of Practice Exercises, but only as tips for how things could be made more idiomatic, referring back to the Concepts previously taught. We expect this analysis to take lower priority than on Concept Exercises (as students are not blocked) so it should be seen as a nice bonus feature, not something that the exercise or student should rely on.

Once submitted, the student will have the option to browse different "Approaches". These will be articles written by our community explaining the various common (or possible) approaches to this exercises, outlining the pros/cons and tradeoffs of the approach. We will link to curated examples of user-submitted solutions that adhere to the various approaches for students to explore further. We will crowd-source these approaches, giving the community the option to become Exercism Authors, but will have a review process to ensure a solid quality-barrier. For exercises that do not yet have "Approaches" written, we will automatically group solutions by AST analysis, for students to browse.

Finally, students can ask the community a question about their exercise. e.g. "I've read this approach but I don't understand how XYZ works and how I can apply it to my code". Community members can then reply to the student with their thoughts - a little like the community interactions on v1. We will make it explicit to students that when they ask a question they are not getting "mentor-quality" feedback, but tapping into the hive-mind of the community. This will also help us add a fair gatekeeping level for mentors. Community-mentors who answer questions helpfully, and have therefore proven their ability to communicate empathetically and helpfully, will have the option to become mentors.

We intend for Practice Exercises to be shared across tracks, as they currently are, but we need to put together solutions for some of the challenges the Problem Specifications repository has encountered recently. We will be writing up a separate issue for discussion about that in the next few days.

### Practice Mode

We also intend to maintain Practice Mode ([formally Independent Mode](https://exercism.io/blog/independent-mode-becomes-practice-mode)), but encourage it as a space for practicing a language that a student knows, rather than learning a new language. It will contain only Practice Exercises and we will make the prerequisite topics on each exercise clearer for those who want to practice certain concepts. Students will be able to ask questions to the community about their exercises, but not directly to mentors.

## Making this a reality

Creating these new type of tracks is an ambitious exercise and one that will rely on our community to work together to make it a reality. I see the primary job of the Exercism Leadership and Staff as empowering our community with the tools they need to build out these new exercises, and to facilitate the discussions required to share and develop knowledge. We will need maintainers to get onboard with these changes and be able to commit the time and energy to develop these new tracks. In this shift, the role of the maintainer is changing from being a person who implements exercises to a person who designs exercises based on their knowledge of the learning journey of their language. Some maintainers may not enjoy that as much, others may thrive on it. We want to support our maintainers by actively recruiting more help for them where there is a shortage, reaching out to established mentors, and the wider community, asking for support. In time we'll be asking each track to assess their capacity to build these new tracks, and to tell us what they need from us to make it happen.

There is also a big chunk of development work to be done on the website-side to implement the changes - both within the backend and the UX. We are targeting an Q1 2019 release of the first v3 tracks, giving us about 5 months to build the changes into the website, and into the tracks themselves.

Once we have had some time to hear everyone's thoughts, and to refine this plan, we will formally write it up into a blog post for the community, and also start to work out guidance for tracks on the most efficient process to build out these tracks.

**Comment by coriolinus** [comment]
- I like this a lot. The biggest failure point of v2 was that it bottlenecked student progress on human analysis. This looks like a good way to avoid that bottleneck. 

- > unlocking will be implicitly worked out by the tagging of "concepts learnt" on Concept Exercises, and "prerequisite concepts" on Practice Exercises

    This is great, but we should standardize on a single term. I suggest "prerequisite concepts". 

- Can we expect a straightforward mapping of "topics" to "concepts" per exercise, at least as a starting point?

**Comment by ErikSchierboom** [comment]
So one thing that's not mentioned but where I do think the new Concept Exercises can be different than the current Core Exercises, is that we can have them build on each other more. When designing Concept Exercises, we can take into account relative difficulty, encountered types and the progression within concepts (starting with basic iteration, moving on to more complex iteration, etc.). These are all concepts that are core to the Track Anatomy tool, and can be extremely useful for a maintainer to help design a track's structure.

**Comment by ErikSchierboom** [comment]
"Tracks implementing an exercise on enumeration can then include this common paragraph, before explaining their own language-specific semantics and syntax."

Is this something we should be teaching, or should the students find out by themselves? So take the example of enumeration. Suppose we are designing an exercise that requires the student to do some basic iteration. In C#, there are several ways one can do iteration (LINQ, foreach, for), which all are valid (but not equally idiomatic in all cases). Wouldn't it be nice if we didn't present the student with these different syntaxes, but let the students find out themselves?

**Comment by Stargator** [comment]
Sent these comments to ihid, figured I post them here:

~A couple of~ Several thoughts:
1) Who is going to create the concept exercises for a given track? If primarily the maintainers, then some tracks may have an easier time collaborating. Dart and Java are very similar, as are many languages.

2) Some overall guidance for the concept exercises and the order of them would be very helpful. I'm not a teacher, so knowing where to start and what could be built on top of it, would be good.

3) For concept exercises, it seems they would start with the assumption that the student doesn't know any language, correct?

4) Regarding Practice Mode, if Concept exercises teach certain topics, then as a user I may find those exercises more interesting to play with than one where a more complex solution that involves more than the concept of interest.

5) I think you mean to say the initial rollout will be Q1 of 2020, Q1 of 2019 already passed, right?

6) As far as implementation, I recommend having both v2 and v3 on the site. So students can wrap up and address current feedback within a publicly defined deadline.

7) To that end from #6, tracks may have four folders `exercises` for v2, and `concept_specs`, `concept_exercises` and `practice_exercises` for v3.

8) As a maintainer, I would like the format of problem specs for the practice exercises to be similar to the ones currently defined in problem-specifications. My current thought, would be to leverage current tooling so the track can generate exercises for practice and concepts.

**Comment by kotp** [comment]
>  Q1 2019

The Q1 2019 release for V3 is rather ambitious @iHiD  ;P

**Comment by tehsphinx** [comment]
I think this is very good! The focus on the concepts of a language makes me 😄 !

**Comment by cmccandless** [comment]
RE @Stargator
> For concept exercises, it seems they would start with the assumption that the student doesn't know any language, correct?

I would argue against that assumption. Exercism is aimed at programmers who know the basics (data types, flow control, OOP, etc), but wish to learn them in a new language. So while the student wouldn't need to know a *specific* language to start a new track, it is a fair assumption that they are not completely new to programming.

**Comment by iHiD** [comment]
Thanks for all the comments so far, everyone. I've tried to address any questions asked below. We'll take the other (non-question) feedback all into account as we move forward. If I've missed anything, please comment again and I'll answer! :)

---

> Can we expect a straightforward mapping of "topics" to "concepts" per exercise, at least as a starting point?

We have a project that we're starting to map out all programming concepts and how they map onto languages. @ErikSchierboom is heading that up and we'll be writing about it in the next couple of weeks and opening up a repo with some initial scaffolding work for people to base things from. I don't know enough about the topics to say whether that will be transferable, but as Erik is au fait with them, I'm sure he can mention that when he writes up an issue.

---

> Is this something we should be teaching, or should the students find out by themselves? So take the example of enumeration. Suppose we are designing an exercise that requires the student to do some basic iteration. In C#, there are several ways one can do iteration (LINQ, foreach, for), which all are valid (but not equally idiomatic in all cases). Wouldn't it be nice if we didn't present the student with these different syntaxes, but let the students find out themselves?

I feel like we need to experiment with some examples to get the balance - I don't know what that is until we achieve it. My initial instinct says that, as someone moving to C#, I'd like the instructions to tell me that `LINQ`, `for` and `foreach` exist, and give me links to them, but then expect me to work out the exact syntax/usage to solve the exercise. So the exercise might be to do something that can be done in any of those three ways, then the post-solution comments explain why one is better, or the exercise might encourage the student to use all three somehow. We're going to need to experiment and play to work it out.

---

> Who is going to create the concept exercises for a given track? If primarily the maintainers, then some tracks may have an easier time collaborating. Dart and Java are very similar, as are many languages.

Maintainers. Collaboration is going to be key. I hope we can build language-specific teams, and also a team that has a focus/skill in creating great exercises generally. Working out how we can keep things lang-specific, but also share knowledge, is going to be the deciding factor in making this works.

---

> Some overall guidance for the concept exercises and the order of them would be very helpful. I'm not a teacher, so knowing where to start and what could be built on top of it, would be good.

Definitely. Having a good set of examples to start with will be key, and then having people who really "get" this to help other maintainers who are working it out for their tracks is going to be an important thing to embed in Exercism generally.

---

> For concept exercises, it seems they would start with the assumption that the student doesn't know any language, correct?

No. The tracks will all be presuming that the users can program, and know at least one language to a vaguely competent standard, ie they are familiar with the basic programming concepts and types in that language. We are planning on building a total-beginner onroad to exercism, but only once this whole bit works.

---

>  Regarding Practice Mode, if Concept exercises teach certain topics, then as a user I may find those exercises more interesting to play with than one where a more complex solution that involves more than the concept of interest.

Yep - that's true. And maybe we introduce that eventually. However, as I generally have a focus on doing the absolute simplest thing for Practice Mode (because of time/energy) I think just giving the traditional set of exercises will be best for now.

---

> I think you mean to say the initial rollout will be Q1 of 2020, Q1 of 2019 already passed, right?

Yes. Thank you @kotp for fixing this.

---

> As far as implementation, I recommend having both v2 and v3 on the site. So students can wrap up and address current feedback within a publicly defined deadline.

I imagine we will have to do this simply because based on the v1-v2 experience, I expect many tracks won't be anywhere near ready, due to resource. How exactly we do this, and communicate it to everyone, will be a challenge we need to solve.

---

> To that end from number 6, tracks may have four folders exercises for v2, and concept_specs, concept_exercises and practice_exercises for v3.

Possibly. I have no idea about this bit yet.

---

> As a maintainer, I would like the format of problem specs for the practice exercises to be similar to the ones currently defined in problem-specifications. My current thought, would be to leverage current tooling so the track can generate exercises for practice and concepts.

Practice exercises will be grounded in problem-specifications still, but problem-specifications will undergo some changes.

**Comment by ErikSchierboom** [comment]
> Can we expect a straightforward mapping of "topics" to "concepts" per exercise, at least as a starting point?

Well, potentially: yes. That said, I'm not sure the concepts assigned to exercises are in all cases the most representative concepts. So this is something that we'll have to look at, but some initial mapping could very well be useful.

**Comment by pgaspar** [comment]
Great stuff 👍 I think the problems you identified for V2 are spot on. I'm still digesting the V3 approach, trying to understand the different contributors we'll need to make this work.

- Mentors: since Practice Exercises will not be mentored and 95% of Concept Exercises will be handled by auto-analysis, will mentors only be responsible for mentoring Concept Exercises that were not auto-analyzed?
- Maintainers: will be responsible for designing and structuring Concept Exercises and how they unlock Practice Exercises.
- Community members: will be able to post and reply to exercises questions.
- Exercism Authors: will be responsible for writing up Approaches for Practice Exercises.

Is this right?

**Comment by cmccandless** [comment]
> Mentors: since Practice Exercises will not be mentored and 95% of Concept Exercises will be handled by auto-analysis, will mentors only be responsible for mentoring Concept Exercises that were not auto-analyzed?

I believe that every Concept Exercise is meant to be auto-analyzed, but as is the case now, sometimes a solution will fail auto-analysis but need mentor input for the next step. That is where the 5% comes in.

>  Maintainers: will be responsible for designing and structuring Concept Exercises and how they unlock Practice Exercises. Are maintainers also responsible for writing language-specific instructions explaining semantics, syntax, etc. for the Concept Exercises or will that be someone else's responsibility? Like the way the website-copy team wrote the mentor notes.

(@ErikSchierboom, please correct me if I get this wrong) It will be on the student to do their research to learn new syntax/semantics for a language. The Concept Exercise will simply validate that they have learned the new concept, it won't teach it to them.

**Comment by coriolinus** [comment]
This sounds like the v3 structure is likely to end up being very wide and
flat: each successful concept is likely to unlock a medium number of new
concept exercises and a large number of practice exercises. Is that more or
less correct?

**Comment by ErikSchierboom** [comment]
> since Practice Exercises will not be mentored and 95% of Concept Exercises will be handled by auto-analysis, will mentors only be responsible for mentoring Concept Exercises that were not auto-analyzed?

Initially, I think this will be true. @iHiD?

> Community members: will be able to post and reply to exercises questions.

Correct.

> Exercism Authors: will be responsible for writing up Approaches for Practice Exercises.

Correct. This will be a community-wide effort.

> Maintainers: will be responsible for designing and structuring Concept Exercises and how they unlock Practice Exercises

> Are maintainers also responsible for writing language-specific instructions explaining semantics, syntax, etc. for the Concept Exercises or will that be someone else's responsibility? Like the way the website-copy team wrote the mentor notes.

I think that these instructions are fairly similar to what mentoring notes should achieve, so I think it will be the responsibility of the website-copy team (although maintainers can of course _also_ be in the website-copy team).

> The Concept Exercise will simply validate that they have learned the new concept, it won't teach it to them.

This is still up for debate, but the way you describe it aligns with how I'm currently thinking we should be teaching concepts. There will be some trial and error here, to see what the best approach is.

> This sounds like the v3 structure is likely to end up being very wide and
flat: each successful concept is likely to unlock a medium number of new
concept exercises and a large number of practice exercises. Is that more or
less correct?

The v3 structure is definitely less linear, in that it allows for greater branching. This doesn't mean that we'll open up _all_ related exercises. We still have lots of work to do in this department, but I could imagine that some exercises would require multiple exercises as prequisites. This is, again, something that requires more thought.

**Comment by coriolinus** [comment]
>  some exercises would require multiple exercises as prerequisites

I like that idea. It allows, for example, Rust to enforce a reasonable
ordering between Luhn, Luhn-Trait, and Luhn-From. Still, in the general
case, it's probably a good idea if most exercises' list of prerequisites
contains only concepts.

**Comment by Stargator** [comment]
> some exercises would require multiple exercises as prequisites

But only in the sense of the concepts within an exercise, not making an exercise a prerequisite, but the concepts within it.

**Comment by ErikSchierboom** [comment]
> But only in the sense of the concepts within an exercise, not making an exercise a prerequisite, but the concepts within it.

Yes, that is the idea.

**Comment by iHiD** [comment]
>Are maintainers also responsible for writing language-specific instructions explaining semantics, syntax, etc. for the Concept Exercises or will that be someone else's responsibility? Like the way the website-copy team wrote the mentor notes.

Good question. I want concept exercises to teach enough of a concept that someone can experiment with it to solve the exercise. 

I see tracks being "run" by a wider group of people than "maintainers" who are responsible for implementing things. They need people to design the track structure, people to create (in code) the exercises, people to write the readmes, people to give the feedback, people to create an analyzer, people to create a test-runner etc. Different people will enjoy different bits of that and be good at different bits of that. We'll need to think about how we structure things to make this work, but I definitely see us having a "Ruby team" who are responsible for making the Ruby track awesome. I imagine that the initial members of that team will be existing maintainers, website-copy team, and top maintainers. For some tracks this might be 1 or 2 people, for others it might be 10pps. 

Over the next few months, I see a big part of my role in facilitating the human side of all this, and @ErikSchierboom's role in facilitating the technical/structural side of this. But it's all to me worked out / firmed up, and thoughts are welcome :)

**Comment by iampeterbanjo** [comment]
TL;DR - Please don't block student progress - I'm blocked right now and it's frustrating. :(

Hi, thanks for creating this project and opening the discussion. I use exercism as a level-up tool to get some deliberate practice with a new language. Today, I experimented with the mentor track learning Elixir and I'm blocked from progressing and there is no way to switch to independent mode. Why is that? I'm an experienced programmer, open to mentorship but happy to progress at my own pace.

My feeling of the discussion (please correct me if I've misjudged) is that there is a tension between a top-down learning style e.g. autoanalysis, mentors, learning tracks etc. and a self-paced learning e.g. do whatever exercise whenever. I would favour a self-paced approach with **some** top-down special sauce a lot like Stackoverflow. Experts will arise via their activity which is publicly visible, students can help curate content via ratings, tags, etc. and some automation can be done to move things along e.g. tags, ratings, moderation etc.

So I'm loving this:

> They need people to design the track structure, people to create (in code) the exercises, people to write the readmes, people to give the feedback, people to create an analyzer, people to create a test-runner etc. Different people will enjoy different bits of that and be good at different bits of that. We'll need to think about how we structure things to make this work, but I definitely see us having a "Ruby team" who are responsible for making the Ruby track awesome.

But not so much this:

> We will also change the way unlocking works, so rather than being explicitly stated in track configurations, unlocking will be implicitly worked out by the tagging of "concepts learnt" on Concept Exercises, and "prerequisite concepts" on Practice Exercises. Once enough Concept Exercises have been completed that all the prerequisite concepts for a Practice Exercise has been been fulfilled, that Practice Exercise will be unlocked.

My argument to support my position is that any rule around "unlocking" a feature/track/lesson/exercise etc. contains an assumption that is open to challenge (e.g. prerequisites assume I'm new to the language or I only use exercism) and blocks progress. Please don't block - let's be async ;)

**Comment by cmccandless** [comment]
> and there is no way to switch to independent mode.

There is. There's a button "Switch to Practice Mode" under "Danger Zone" at the bottom of the right column on the track page.

**Comment by iHiD** [comment]
@iampeterbanjo Thanks for your comment :)

The issues with blocking students right now are because we have to wait for mentors to respond to a student to allow them to progress, which can take a while, and can be frustrating if your code is perfect. The work with analyzers fixes this by meaning that if you submit a solution that is good, you'll automatically be allowed to progress. So under this proposal, if you're comfortable with the basics of a language and want to get to the more complex concept exercises, you should be able to quickly bash through the first few exercises and have them automatically approved.

The changes to unlocking are changes to how things are unlocked (looking at concepts rather than manual crafting in the config.json), and don't effect the general mechanisms of unlocking or progression in the way that (I think) is frustrating for you.

As @cmccandless says, you should be able to switch to Practice Mode from the track page so that you don't have to wait any more. Feel free to open a separate issue if you don't see that button and someone can help with that :)

**Comment by MarkusReynolds1989** [comment]
This an absolutely brilliant idea. Take for instance recently I tried to start the c track, I'm a dotnet developer, the mentor did their best but I still couldn't understand memory management 100 percent. It'd be extremely helpful if for the exercise there could be a lesson related to it and then have it relate to an implementation. I also mentor c# and see students struggle with c# specific implementation to solve problems where they rely more on generic programming ( arrays vs collections, linq vs for next loops and state). If there's anything I can do to help I will and I'll keep track of this page.

**Comment by iampeterbanjo** [comment]
Thank you @iHiD for your response. :)

@cmccandless - thanks for pointing out the "Practice mode" button - I missed that.

@MarkusReynolds1989 +1 on having a lesson and then a practice exercise. Actually, how are the exercisms decided at the moment?

I think where I differ is that there is a lock - I'm suggesting we have no locks but can invite mentors for code review. It would also be less taxing on mentors because they'll be going where they're wanted. I don't see why it's "have a mentor and have locks" or "have no mentor and have no locks", how about "invite mentor and have no locks"? 

Another aid to a self-paced learning style would be to give or highlight a model answer in the list of solutions.

**Comment by iHiD** [comment]
> I think where I differ is that there is a lock - I'm suggesting we have no locks but can invite mentors for code review.

The reason for blocking progress is that without it students end up on complex exercises without having actually understood the pre-reqs, which then puts a huge amount of weight on later mentors who are having to teach way too much on a single exercise. You can read more about top mentors thoughts on this [here](https://github.com/exercism/exercism/issues/4908). 

It's also worth me saying that I see very little chance of this changing. I've spoken to a lot of mentors to understand what they think and the product team see progression as an important component of the Exercism experience working. I agree that it is currently problematic, but I think that is caused by other things (such as the student/mentor quantity mismatch), and that this issue is moving us to a point where if you submit a solution that is approvable, it will automatically be approved by the analyzers in the vast majority of cases, which would remove the problem.

I'm happy to discuss this further on a different issue, but as this issue doesn't change progression-blocking behaviour, I don't think this is the right place to continue the discussion.

**Comment by MarkusReynolds1989** [comment]
> > I think where I differ is that there is a lock - I'm suggesting we have no locks but can invite mentors for code review.
> 
> The reason for blocking progress is that without it students end up on complex exercises without having actually understood the pre-reqs, which then puts a huge amount of weight on later mentors who are having to teach way too much on a single exercise. You can read more about top mentors thoughts on this [here](https://github.com/exercism/exercism/issues/4908).
> 
> It's also worth me saying that I see very little chance of this changing. I've spoken to a lot of mentors to understand what they think and the product team see progression as an important component of the Exercism experience working. I agree that it is currently problematic, but I think that is caused by other things (such as the student/mentor quantity mismatch), and that this issue is moving us to a point where if you submit a solution that is approvable, it will automatically be approved by the analyzers in the vast majority of cases, which would remove the problem.
> 
> I'm happy to discuss this further on a different issue, but as this issue doesn't change progression-blocking behaviour, I don't think this is the right place to continue the discussion.

I think there has to be a sense of progression to keep things interesting. Back to my C lang example from above, if I was able to get all the tests to pass but one and I was bored of the problem I might move on without solving it. One thing I'd like to see added is some sort of scoring system where the best written and the best performing solutions are more rewarded. I think that'd be pretty tough to implement though, so the focus on "Here's a lesson, here's the scope of the problem, go solve it until you get it right" would be the most ideal for just learning a new language.
As I said, I'm currently blocked in that C track because I don't understand the memory management needed to advance (my tests passed but the mentor wants me to understand the memory management which I think he's right in doing so) but there's no way for me to really understand exactly what to do without learning all of C memory management instead of just the few things I need to complete the lesson.
I. Lesson 1 in C Memory Management part 1
II. Lesson 2 in C Recursion part 1
III. Lesson 3 in C Memory Management part 2
and so on, I think this would be much better and you'd get a much better grasp of what you understand and what you still need to learn that's specific to the language.

**Comment by sshine** [comment]
@MarkusReynolds1989 wrote:
> some sort of scoring system where the best written and the best performing solutions are more rewarded

The only public reward you currently get is your solution in "community solutions" and in your profile.

"Community solutions" will eventually feature a better scoring / grouping mechanism, and whether a solution fails tests or gets flagged by the automated feedback system are good ideas for scoring parameters. There are other more important parameters, such as likeness to other community solutions, so we don't get flooded by basically the same solution.

This is being addressed as a separate issue from v3 tracks, but it's on the drawing board.

> I'm currently blocked in that C track because I don't understand the memory management needed to advance (my tests passed but the mentor wants me to understand the memory management which I think he's right in doing so) but there's no way for me to really understand exactly what to do without learning all of C memory management instead of just the few things I need to complete the lesson.

This sounds like highly valuable advice for bringing the C track into an even better place. There is technically nothing that prevents you from contributing to this iterative improvement of the track today. @exercism/c / @wolf99 may prefer that this happens at https://github.com/exercism/c/issues

> Recursion *[...]* Memory Management

Thanks for pointing out that recursion and dynamic memory management play together. I'll pass that note on to the appropriate place so other languages with memory management may benefit in the v3 track makeover.

**Comment by iampeterbanjo** [comment]
@iHiD thanks for the link to that thread. So it looks like everything I have to say has been said. What I'm taking away from that is there **isn't a platform to get this kind of mentoring** and I have actually been looking but it's not why I joined Exercism - I came for the code problems with solutions. I still wish it didn't block. Lol, :smile: 

Ok, ok, maybe we can't agree 100% but I'll play along. My mentor is trying their best and it seems to be a 12 hour wait time which is reasonable. It could be we just need some UX touches like projected wait times, easy switching, mentor profiles, required reading etc.

Thanks for all the hardwork and hearing me out.

**Comment by wolf99** [comment]
@MarkusReynolds1989 Yes, if you have suggestions or questions on the C track structure, progression or information please do jump on over to the exercism/c track repo and add an issue (or check out the existing ones, we already have a bit of work ongoing to try to improve the track). We''ll be happy to see you there! 😄

**Comment by coriolinus** [comment]
In fairness, I've been in Practice mode on Rust since the switch to v2.
I've also had requests for mentorship queued up since day 2 of v2, which
have never been mentored. Practice mode may no longer be called independent
mode, but in practice, the student is independent.

**Comment by mpizenberg** [comment]
> shoehorning these into a "curriculum" is a painful, and probably impossible, task for maintainers

From my experience in the elm track, I wasn't under the impression that there is curriculum indeed, just a bunch of exercises, with different levels of difficulty. That's also what I expected a site called "Exercism" to be so a "curriculum" comes a bit as a surprise.

> There are also huge waits between submitting an exercise and receiving feedback, and that break of flow reduces a lot of the positive excitement from learning something

That seems to be the case here.

> I'm not convinced that at the end of an exercise a student really understands what they've learnt, or how to apply it in the future.

Made me think of this presentation: "How to teach programming (and other things)?" by Felienne Hermans, https://youtu.be/g1ib43q3uXQ

> Concepts Exercises

Is there a list of those concepts somewhere? Is this the list of "topics"? I can't find the link to where this list exist anymore.
-> Ok I actually found an answer by continuing reading post interactions.

> However, we will write language-agnostic introductions to programming concepts, with links to language-generic resources on that topic, which we will ask tracks to use. For example, we might define a concept of "Enumeration" and provide a short introduction that explains how enumeration differs from loops with a link out to a good article (either internal to Exercism or elsewhere on the web).

I am a bit doubtful about the feasibility of these accross paradigms where the differences are not restricted to syntax. I'm still curious to see what you'll come up with. There might be a big enough subset of concepts independants from paradigms.

> Once enough Concept Exercises have been completed that all the prerequisite concepts for a Practice Exercise has been been fulfilled, that Practice Exercise will be unlocked.

This path seems great for beginners, but for more advanced programmers curious about some exercises, it may feel quite annoying. I feel like people should know when to get back at related concept exercises if they are clearly indicated, without the need for a hard constraint. But maybe I'm wrong, or maybe it's just subjective.

**Comment by kfedorov91** [comment]
> Secondly, in our strategies in trying to achieve this, we have focused too much on what was similar about languages, and not looked enough at the key question of "what makes languages unique?"

Strongly agree with this. I'm trying to learn Rust right now and Rust has so much cool stuff in its standard library that can be totally missed by a newbie grinding out algorithm-based questions on generic online judge sites like LeetCode/HackerRank/CodeWars/etc because 95%+ of the questions can be solved just using functions, arrays, and for-loops so those sites don't really teach anything about languages other than the lowest common denominator of features between all of them, which is a pretty low bar to hit.

> At the moment the journey for a student is:
> - Be given an exercise.
> - Try something - normally via an incorrect approach grounded from previous language experience.
> - Be told what you should have been doing.
> - Work out how to redo it that way.

Speaking as a student who signed up a few days ago to join the Rust track my journey so far has been:
- Be given an exercise.
- Try something - normally via an incorrect approach grounded from previous language experience.
- Wait for mentor feedback.
- Keep waiting...
- "There's still 160 exercises ahead of me in the queue!? But it was 170 over 2 days ago! And this track has 38 mentors? How!? Are any of them still active?"
- Lose faith in the mentorship system and switch to practice mode.
- Try to figure out what idiomatic code for the language looks like from the most starred community solutions.

I have some suggestions for improvement for the existing v2 system:

1) The queue prioritizes core exercises over side exercises and mentored mode users over practice mode users. Great, however I think it would be even better if the queue also prioritized users who are completely blocked (mentored mode + finished all available side exercises) over users who are partially blocked (mentored mode + still have available side exercises). Actually, the former rule can be generalized to: users who have completed more exercises should have higher priority in the queue than users who have completed less exercises. This rule not only captures users in mentored mode who have completed all the available side exercises but it also captures users who have gotten so frustrated with waiting around in mentored mode that they have switched to practice mode to move on ahead in the track. A practice mode user who has completed 30 exercises in a track should get more priority than a mentored mode user who has only completed 3 exercises because the former has shown a great deal of eagerness and commitment to learning the material and should get rewarded for that. I think the rule can extend to completed exercises even outside of the current track: a practice mode user in track A who's a long-time Exercism user and who has completed 203 exercises already across tracks A, B, and C should get more priority than a mentored mode user in track A who is brand new to the site and who has only completed 3 exercises. At that point you don't even need the concepts of a "practice" or "mentored" mode and can go by completed exercises.
2) Mentors need to hit a minimum weekly quota to stay mentors. A mentor who does zero mentorship is not a mentor. Exercism is a fairly well-known and popular site and I imagine being able to say you're a mentor on Exercism in your Github or Twitter bio comes with a certain amount of nerd cred, but that nerd cred needs to have some kind of maintenance cost. Despite what Exercism claims the Rust track most definitely does not currently have 38 active mentors, maybe 2-3 tops. Even if the minimum quota was super low, like reviewing just 3 exercises per week, the entire Rust queue would be empty by the end of next week.
3) Mentors for a track should get pinged by email or phone if the queue gets too large (if they don't already), and the pings should be persist and annoying and they can't opt-out without also stepping down from their mentor position.
4) Mentors who go above and beyond the minimum quota should get more recognition and nerd cred. The top 10-20% should be "Gold Star Mentors" or "Top Mentors" or whatever you want to call it and it should be displayed prominently in their profiles.
5) If all else fails users should be able to pay a small amount of money to get put further ahead in the queue, and mentors who provide them feedback should get some share of that money.

Everything else you wrote regarding the v3 system sounds awesome though and I can't wait to see it.

**Comment by iHiD** [comment]
Thanks everyone. Lots of interesting thoughts! 

As everyone seems to be agreeing that these new tracks sound (at least in general) good, I'm going to close this issue, and continue to work on this with the product team. We'll take all the ideas here on board as part of that.

We'll also be announcing more details on "concepts" side of things in the next week or so, which we can discuss more on a separate issue then.

For anyone interested in where this fits into the general Exercism roadmap, don't forget our [strategy document](https://exercism.io/strategy) which outlines our general direction, and [this blog post](https://exercism.io/blog/whats-next-for-exercism-aug-2019) which outlines things a bit further.
